design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+      spec:
+        forProvider:
+          databaseVersion: POSTGRES_9_6
+          region: us-central1
+          settings:
+            tier: db-custom-1-3840
+            dataDiskType: PD_SSD
+    patches:
+    - type: FromCompositeFieldPath
+      fromFieldPath: spec.parameters.storageGB
+      toFieldPath: spec.forProvider.settings.dataDiskSizeGb
+```
+
+Limiting the functionality of our `Composition` type allows us to reduce
+Crossplane's learning curve and implementation complexity. At the same time it
+limits Crossplane's ability to address many use cases. It's not uncommon for
+Crossplane users to resort to 'composing' Crossplane managed resources using
+more featureful tools such as Helm or jsonnet when they hit the limitations
+imposed by the `Composition` type, or even writing their own Kubernetes
+controllers (typically in Go, using [controller-runtime]) to build bespoke
+abstractions. Furthermore, some Crossplane users simply prefer to express their
+intent in a language (or with a tool) they already know and feel comfortable
+with, such as Helm, jsonnet, Python, HCL, Typescript, etc.
+
+While it's true that users _could_ simply use Helm, jsonnet, etc etc to compose
+Crossplane managed resources into higher level abstractions (like the
+aforementioned `AcmeCoDatabase` example), we feel that these abstractions are
+best exposed as _APIs_, not as client-side constructs.
+
+Exposing abstractions as a (Kubernetes) API:
+
+* Avoids the need to distribute (and version) client-side tooling as part of the
+  collaboration process.
+* Allows abstractions to be access controlled using RBAC.
+* Increases tooling support - more tools have native support for calling REST
+  APIs than (e.g.) building Helm charts.
+* _Further_ increases tooling support by leveraging the Kubernetes ecosystem -
+  things like ArgoCD and OPA.
+
+## Goals
+
+The proposal put forward by this document should:
+
+* Let folks use their composition tool and/or programming language of choice.
+* Support 'advanced' composition logic such as loops and conditionals.
+* Balance safety (e.g. sandboxing) with speed and simplicity.
+* Be possible to introduce behind a feature flag that is off by default.
+
+While not an explicit goal, it would also be ideal if the solution put forth by
+this document could serve as a test bed for new features in the contemporary
+'resources array' based form of Composition.
+
+## Proposal
+
+### Overview
+
+This document proposes that a new `functions` array be added to the existing
+`Composition` type. This array of functions would be called either instead of or
+in addition to the existing `resources` array in order to determine how an XR
+should be composed.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: my-cool-function
+    type: Container
```

Comment:
This implies the Composition owns the function lifecycle and I really get scared when I see this much coupling. Why not rely on references to the function resources? For example, you can have a reference to a Kubernetes Service which is pointing to a running function, if the Service is referenced in the composition is not available, then the composition is not ready to be used.

Reply:
What you're describing sounds close to the webhook function variant that we envision adding at some point in the future. In the meantime I think it's important to support OCI based functions because they're dramatically easier to build and deploy than spinning up something that can serve a webhook (e.g. Deployment + Service + Certs etc).

> I really get scared when I see this much coupling.

Why?